tests/test_ag_ui.py

```diff
@@ -0,0 +1,941 @@
+    """Example state class for testing purposes."""
+
+    value: int = 0
+
+
+def get_weather(name: str = 'get_weather') -> Tool:
+    return Tool(
+        name=name,
+        description='Get the weather for a given location',
+        parameters={
+            'type': 'object',
+            'properties': {
+                'location': {
+                    'type': 'string',
+                    'description': 'The location to get the weather for',
+                },
+            },
+            'required': ['location'],
+        },
+    )
+
+
+@pytest.fixture
+async def adapter() -> Adapter[StateDeps[StateInt], str]:
+    """Fixture to create an Adapter instance for testing.
+
+    Returns:
+        An Adapter instance configured for testing.
+    """
+    return await create_adapter([])
+
+
+async def create_adapter(
+    call_tools: list[str] | Literal['all'],
+) -> Adapter[StateDeps[StateInt], str]:
+    """Create an Adapter instance for testing.
+
+    Args:
+        call_tools: List of tool names to enable, or 'all' for all tools.
+
+    Returns:
+        An Adapter instance configured with the specified tools.
+    """
+    return Adapter(
+        agent=Agent(
+            model=TestModel(
+                call_tools=call_tools,
+                tool_call_deltas={'get_weather_parts', 'current_time'},
+            ),
+            deps_type=StateDeps[StateInt],  # type: ignore[reportUnknownArgumentType]
+            tools=[send_snapshot, send_custom, current_time],
+        ),
+    )
+
+
+@pytest.fixture
+def mock_uuid(monkeypatch: pytest.MonkeyPatch) -> _MockUUID:
+    """Mock UUID generation for consistent test results.
+
+    This fixture replaces the uuid.uuid4 function with a mock that generates
+    sequential UUIDs for testing purposes. This ensures that UUIDs are
+    predictable and consistent across test runs.
+
+    Args:
+        monkeypatch: The pytest monkeypatch fixture to modify uuid.uuid4.
+
+    Returns:
+        A function that generates a mock UUID.
+    """
+    counter = count(1)
+
+    def _fake_uuid() -> str:
+        """Generate a fake UUID string with sequential numbering.
+
+        Returns:
+            A fake UUID string in the format '00000000-0000-0000-0000-{counter:012d}'.
+        """
+        return f'00000000-0000-0000-0000-{next(counter):012d}'
+
+    def _fake_uuid4() -> uuid.UUID:
+        """Generate a fake UUID object using the fake UUID string.
+
+        Returns:
+            A UUID object created from the fake UUID string.
+        """
+        return uuid.UUID(_fake_uuid())
+
+    # Due to how ToolCallPart uses generate_tool_call_id with field default_factory,
+    # we have to patch uuid.uuid4 directly instead of the generate function. This
+    # also covers how we generate messages IDs.
+    monkeypatch.setattr('uuid.uuid4', _fake_uuid4)
+
+    return _fake_uuid
+
+
+def assert_events(events: list[str], expected_events: list[str], *, loose: bool = False) -> None:
```

Comment:
Can we make this take a list of dicts instead of JSON strings, and then swap the UUIDs in the `expected_events` dicts for `'messageId': IsStr()`, which correctly handles `==` comparison against any string?